### PR TITLE
Set config section keeper/keeper_server  an alternative of zookeeper

### LINKED ISSUE #34766: Recognize `<keeper>` as alternative to `<zookeeper>` in server's config.
**Use case**

Avoid confusion when people are using `clickhouse-keeper`.

**Describe the solution you'd like**

`keeper` section will be identical to `zookeeper`, but don't allow both sections to be present.

### PR DESCRIPTION
**Pull request by taiyang-li**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support config sections `keeper`/`keeper_server` as an alternative to `zookeeper`. Close https://github.com/ClickHouse/ClickHouse/issues/34766 , https://github.com/ClickHouse/ClickHouse/issues/34767

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Maybe better to translate `zookeeper` into `keeper` (or vice versa) during merging configs into preprocessed config?

**Comment by taiyang-li** [comment]
@nikitamikhaylov Good idea, that will reduce a lot of if-elses.

**Comment by alexey-milovidov** [comment]
@antonio2368 The idea of implementation is ok, I think we can merge even without the proposal from @nikitamikhaylov.

File: programs/extract-from-config/ExtractFromConfig.cpp [hunk 1]
```diff
@@ -89,8 +94,14 @@ static std::vector<std::string> extractFromConfig(
     if (has_zk_includes && process_zk_includes)
     {
         DB::ConfigurationPtr bootstrap_configuration(new Poco::Util::XMLConfiguration(config_xml));
-        zkutil::ZooKeeperPtr zookeeper = std::make_shared<zkutil::ZooKeeper>(
-                *bootstrap_configuration, "zookeeper", nullptr);
+
+        if (bootstrap_configuration->has("zookeeper") && bootstrap_configuration->has("keeper"))
+            throw DB::Exception(DB::ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG, "Both zookeeper and keeper are specified");
```
Comment (antonio2368) [inline programs/extract-from-config/ExtractFromConfig.cpp:99]:
```suggestion
            throw DB::Exception(DB::ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG, "Both ZooKeeper and Keeper are specified");
```

File: programs/copier/ClusterCopierApp.cpp [hunk 2]
```diff
@@ -192,6 +197,9 @@ void ClusterCopierApp::mainImpl()
     if (!task_file.empty())
         copier->uploadTaskDescription(task_path, task_file, config().getBool("task-upload-force", false));
 
+    if (config().has("zookeeper") && config().has("keeper"))
+        throw Exception("Both zookeeper and keeper are specified", ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
```
Comment (antonio2368) [inline programs/copier/ClusterCopierApp.cpp:201]:
```suggestion
        throw Exception("Both ZooKeeper and Keeper are specified", ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
```

File: programs/server/Server.cpp [hunk 3]
```diff
@@ -774,7 +774,10 @@ int Server::main(const std::vector<std::string> & /*args*/)
 
     ConnectionCollector::init(global_context, config().getUInt("max_threads_for_connection_collector", 10));
 
-    bool has_zookeeper = config().has("zookeeper");
+    if (config().has("zookeeper") && config().has("keeper"))
+        throw Exception("Both zookeeper and keeper are specified", ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
```
Comment (antonio2368) [inline programs/server/Server.cpp:778]:
```suggestion
        throw Exception("Both ZooKeeper and Keeper are specified", ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
```

File: src/Interpreters/Context.cpp [hunk 4]
```diff
@@ -2128,7 +2128,12 @@ zkutil::ZooKeeperPtr Context::getZooKeeper() const
 
     const auto & config = shared->zookeeper_config ? *shared->zookeeper_config : getConfigRef();
     if (!shared->zookeeper)
-        shared->zookeeper = std::make_shared<zkutil::ZooKeeper>(config, "zookeeper", getZooKeeperLog());
+        shared->zookeeper = std::make_shared<zkutil::ZooKeeper>(
+            config,
+            config.has("zookeeper")    ? "zookeeper"
+                : config.has("keeper") ? "keeper"
+                                       : "keeper_server",
```
Comment (antonio2368) [inline src/Interpreters/Context.cpp:2135]: Please create a function for this.
You can also store in `std::array`, iterate and return first contained in config.

File: tests/integration/helpers/zookeeper_secure_config.xml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <clickhouse>
-    <zookeeper>
+    <keeper>
```
Comment (antonio2368) [inline tests/integration/helpers/zookeeper_secure_config.xml:2]: test with `keeper_server` is needed.

and don't just change here, create a test which will test 3 different configs each using a different way of defining ZooKeeper cluster.
You create a ReplicatedMergeTree and test that an insert works.

**Comment by antonio2368** [comment]
@taiyang-li any progress here?

**Comment by taiyang-li** [comment]
@antonio2368  Sorry, I'm busy working on other PRs in recent days. I'll modify code as you requested ASAP.

**Comment by antonio2368** [comment]
Okay no rush, just wanted to know if you're okay with making changes. 😄

File: src/Common/ZooKeeper/examples/zk_many_watches_reconnect.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
+#include <Common/Config/ConfigProcessor.h>
```
Comment (antonio2368) [inline src/Common/ZooKeeper/examples/zk_many_watches_reconnect.cpp:1]: This is removed in master, any reason to keep it?

File: src/Common/ZooKeeper/ZooKeeper.cpp [hunk 7]
```diff
@@ -1333,4 +1334,16 @@ String getSequentialNodeName(const String & prefix, UInt64 number)
...
+String getZookeeperConfigName(const Poco::Util::AbstractConfiguration & config)
+{
+    if (config.has("zookeeper"))
+        return "zookeeper";
+    else if (config.has("keeper"))
+        return "keeper";
+    else if (config.has("keeper_server"))
+        return "keeper_server";
+    else
+        throw DB::Exception("There is no Zookeeper configuration in server config", DB::ErrorCodes::NO_ELEMENTS_IN_CONFIG);
```
Comment (antonio2368) [inline src/Common/ZooKeeper/ZooKeeper.cpp:1346]:
```suggestion
    static constexpr std::array zookeeper_config_keys{"zookeeper", "keeper", "keeper_server"};
    for (const auto & key : zookeeper_config_keys)
    {
        if (config.has(key)
            return key;
    }
    throw DB::Exception("There is no Zookeeper configuration in server config", DB::ErrorCodes::NO_ELEMENTS_IN_CONFIG);
```

File: src/Interpreters/Context.cpp [hunk 8]
```diff
@@ -2273,7 +2273,8 @@ bool Context::tryCheckClientConnectionToMyKeeperCluster() const
     try
     {
         /// If our server is part of main Keeper cluster
-        if (checkZooKeeperConfigIsLocal(getConfigRef(), "zookeeper"))
+        if (checkZooKeeperConfigIsLocal(getConfigRef(), "zookeeper") || checkZooKeeperConfigIsLocal(getConfigRef(), "keeper")
```
Comment (antonio2368) [inline src/Interpreters/Context.cpp:2276]: you can get config with `zkutil::getZookeeperConfigName(getConfigRef())`
and this check becomes
`if (config_name == "keeper_server" || checkZooKeeperConfigIsLocal(getConfigRef(), config_name))`

**Comment by alexey-milovidov** [comment]
Let's continue...

**Pull request merged**